runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java

```diff
@@ -56,6 +57,27 @@
 
   private Environments() {}
 
+  public static Optional<Environment> getEnvironment(
+      String ptransformId, Components components) {
+    try {
+      PTransform ptransform = components.getTransformsOrThrow(ptransformId);
+      String envId =
+          KNOWN_URN_SPEC_EXTRACTORS
+              .getOrDefault(ptransform.getSpec().getUrn(), DEFAULT_SPEC_EXTRACTOR)
+              .getEnvironmentId(ptransform);
+      if (!Strings.isNullOrEmpty(envId)) {
+        // Some PTransform payloads may have an empty (default) Environment ID, for example a
+        // WindowIntoPayload with a known WindowFn. Others will never have an Environment ID, such
+        // as a GroupByKeyPayload, and the Default extractor returns null in this case.
+        return Optional.of(components.getEnvironmentsOrThrow(envId));
+      } else {
+        return Optional.empty();
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```

Comment:
It's a little strange to have an exception-throwing method that returns an optional. On the other hand, IOExceptions _really_ shouldn't happen here, so maybe this doesn't matter.

Reply:
I don't really agree with the former claim, though I also agree strongly with the latter.

Optional is the `maybe` type - in the happy case, we always return a thing, even if that thing is the explicit absence of a thing. If we fail for any reason (we got a malformed proto, or whatever), there's not meaningfully a result - it's not unknown, it's a precondition failure (this could probably be `throw new IllegalArgumentException("complaints", e)`, for example).

Reply:
I agree in practice, but only because Java doesn't provide `either` for exceptional outcomes. In any case, it's irrelevant here. 😉